### PR TITLE
theano.tensor.constant reshape fix

### PR DESCRIPTION
**Pull request by SinaHonari**
fixes issue #3031

### CODE REVIEW

File: theano/tensor/basic.py [hunk 1]
```diff
@@ -4126,9 +4126,14 @@ def make_node(self, x, shp):
...
-        if not shp.dtype.startswith('int'):
+        if not shp.dtype.startswith('int') and not(isinstance(shp,
+                                                   TensorConstant) and
+                                                   shp.data.size == 0):
+            # if shp is an instace of TensorConstant and
+            # shp.data is empty, then shp.dtype is float64
             raise TypeError("Shape must be integers", shp, shp.dtype)
-        assert shp.ndim == 1
+        if not(isinstance(shp, TensorConstant) and not shp.data.size == 0):
+            assert shp.ndim == 1
```
Comment (bouthilx) [inline theano/tensor/basic.py]: I was wrong, I verified and ```as_tensor_variable(()).ndim == 1```. You don't need to modify this.

File: theano/tensor/basic.py [hunk 2]
```diff
@@ -4126,9 +4126,14 @@ def make_node(self, x, shp):
         x = as_tensor_variable(x)
         shp_orig = shp
         shp = as_tensor_variable(shp, ndim=1)
-        if not shp.dtype.startswith('int'):
+        if not shp.dtype.startswith('int') and not(isinstance(shp,
+                                                   TensorConstant) and
+                                                   shp.data.size == 0):
```
Comment (bouthilx) [inline theano/tensor/basic.py]: I think it would be more readable this way
```python
if (not shp.dtype.startswith('int') and
    not (isinstance(shp, TensorConstant) and shp.data.size == 0)):
```
And you could even factor out the negation
```python
if not (shp.dtype.startswith('int') or
        (isinstance(shp, TensorConstant) and shp.data.size == 0)):
```

File: theano/tensor/basic.py [hunk 3]
```diff
@@ -4126,9 +4126,14 @@ def make_node(self, x, shp):
...
+            # if shp is an instace of TensorConstant and
```
Comment (bouthilx) [inline theano/tensor/basic.py]: instance*

File: theano/tensor/basic.py [hunk 4]
```diff
@@ -4126,9 +4126,14 @@ def make_node(self, x, shp):
...
+            # shp.data is empty, then shp.dtype is float64
```
Comment (bouthilx) [inline theano/tensor/basic.py]: I would rephrase it as : 
> It raises an error if shp is not of integer type, except when shp is constant and empty (shp.dtype does not matter anymore).

**Comment by bouthilx** [comment]
@SinaHonari 
Thanks for the quick PR!

You still need to add a test case to ensure this bug does not happen again. Add a new method ```test_empty_shp()``` at the end of this [class](https://github.com/Theano/Theano/blob/master/theano/tensor/tests/test_basic.py#L4914)

**Comment by SinaHonari** [comment]
Boss your comments and concerns are all realized :)

**Comment by bouthilx** [comment]
Thanks Sina! But you forgot about the unit-test. ;)

**Comment by SinaHonari** [comment]
What should the unit test do?

**Comment by bouthilx** [comment]
Test the basic case reported and make sure it doesn't fail. #3031 

```python
theano.tensor.constant([1]).reshape(())
```

As I said above

> You still need to add a test case to ensure this bug does not happen again. Add a new method ```test_empty_shp()``` at the end of this [class](https://github.com/Theano/Theano/blob/master/theano/tensor/tests/test_basic.py#L4914).

Should be quite simple. :)

**Comment by SinaHonari** [comment]
You can check if the test is sufficient.

**Comment by bouthilx** [comment]
Please compile the function and verify that output.shape == ()
That will be sufficient. :)

**Comment by bouthilx** [comment]
Looks good to me, Thanks Sina!
